### PR TITLE
strict mode to avoid nested scan

### PR DESCRIPTION
**Pull request by kyunghyuncho**
This addresses the findings from #2609 . 

If strict=True in theano.scan, it assumes that all necessary shared variables in fn are passed as a part of non_sequences.  This has to be ensured by the user. Otherwise, it'll result in an error. This avoids Scan Op calling any earlier (external) Op over and over, and greatly speeds up.

### CODE REVIEW

**Comment by abergeron** [comment]
This looks ok and I have no problem merging it.  However it's still scan code so I would like to have somebody more familiar (like @carriepl) look over it.

**Comment by nouiz** [comment]
Should we wait after Theano 0.7 release to be safe? I do not want to take
risk and introduce new problems.

**Comment by kyunghyuncho** [comment]
I'm okay with waiting, as long as it's not going to be too far (when are
you planning to release 0.7?) Also, note that with the default setting
(strict=False), nothing changes.

**Comment by nouiz** [comment]
Are you sure an error is raised if the user specify strict while the condition isn't there?

Also, can you give some speed up number like you did? Now grad take Xm, while it didn't finished in 30m before this PR?

I just do not see what work this PR remove that would give a significant speed up. @abergeron do you see that?

**Comment by kyunghyuncho** [comment]
On Sun, Mar 15, 2015 at 1:34 PM, Frédéric Bastien <notifications@github.com>
wrote:

> Are you sure an error is raised if the user specify strict while the
> condition isn't there?
>
> Also, can you give some speed up number like you did? Now grad take Xm,
> while it didn't finished in 30m before this PR?
>

It depends heavily on the model. With my code (two scans in parallel
followed by two other scans in sequence, where the first two scan implement
GRU and the other two scan implements GRU with attention mechanism), the
time taken by tensor.grad decreased from roughly several hours to 20
seconds. Similarly, I see the speed-up in doing theano.function returning
the resulting the gradient obtained by tensor.grad.


> I just do not see what work this PR remove that would give a significant
> speed up. @abergeron <https://github.com/abergeron> do you see that?
>
With strict=True, it removed any unnecessary shared variables from
other_shared_scan_args which were added with strict=True because they were
used in the previous scan's (from dummy_f.maker.expanded_inputs). This is a
thing that needs to be further investigated later on why theano.function
returns an instance having shared variables that were used to compute the
"inputs" as a part of maker.expanded_inputs (since "inputs" are given, I
don't think any computation resulted in those "inputs" should be included
in maker.expanded_inputs.)

**Comment by abergeron** [comment]
This is because scan uses a dummy function to grab the shared variables and function actually needs to know all the shared variables implicated in the graph.

We could probably use graph.inputs() instead, but that is too much of a change for me to be comfortable doing right now.

**Comment by nouiz** [comment]
extended_input by definition is the inputs + shared var. This was always
like this and should stay like this. But maybe this isn't what scan need.

if 2 of @carriepl @abergeron or @lamblin tell that it is safe to merge, I'm
ok with this. @abergeron already told this.

I didn't understood this sentense:

"Similarly, I see the speed-up in doing theano.function returning
the resulting the gradient obtained by tensor.grad."

Do you mean this also speed up the compilation?

**Comment by kyunghyuncho** [comment]
@nouiz The speed-up in theano.function(...,tensor.grad(...)) is due to a much simpler graph obtained by tensor.grad. So, it's essentially the same thing helping two functions (tensor.grad and theano.function.)

**Comment by carriepl** [comment]
So, if I'm getting this right, a bad graph with strict=True will only produce an error later when Theano figures out it doesn't have the required inputs for the inner function? If that is indeed the case, the error message will probably not be very useful to the user. It would be best right now that some shared inputs are not in the non_sequences and raise an exception with a meaningful message.

Also, if this offers this kind of speed up, it should also be documented  in the official scan documentation, at least as a *scan best practice*, to pass shared variables explicitly as non-sequences.

File: theano/scan_module/scan.py [hunk 1]
```diff
@@ -314,6 +316,10 @@ def f(x):
         Set the value of allow gc for the internal graph of scan.  If
         set to None, this will use the value of config.scan.allow_gc.
 
+    :param strict:
+        If true, all the shared variables used in ``fn`` must be provided as a
+        part of ``non_sequences``.
```
Comment (lamblin) [inline theano/scan_module/scan.py]: or `sequences`.

**Comment by lamblin** [comment]
I agree this is a useful change, but this PR needs:
- [ ] documentation of the `strict` mode, how to use it, etc., to be put in http://deeplearning.net/software/theano/library/scan.html
- [ ] better error messages when shared inputs are not provided
- [ ] unit tests on which cases are expected to work (and that the result is the same as `strict=False`), and which are supposed to fail (and check that the appropriate error is raised).

**Comment by kyunghyuncho** [comment]
On Mon, Mar 16, 2015 at 4:01 PM, carriepl <notifications@github.com> wrote:

> So, if I'm getting this right, a bad graph with strict=True will only
> produce an error later when Theano figures out it doesn't have the required
> inputs for the inner function? If that is indeed the case, the error
> message will probably not be very useful to the user. It would be best
> right now that some shared inputs are not in the non_sequences and raise an
> exception with a meaningful message.
>
The error message raised currently is quite useful (assuming that every
shared variable was given a name). The error message shows the name of the
shared variable used by the inner function but not supplied as a part of
non_sequences.


> Also, if this offers this kind of speed up, it should also be documented
> in the official scan documentation, at least as a *scan best practice*,
> to pass shared variables explicitly as non-sequences.
>
I can try to document it briefly later when I have more time. But,
meanwhile, I think this should be merged soon to be used by other libraries
on top of Theano (e.g., Blocks.)

File: theano/scan_module/scan.py (hunk 1 above)
Comment (kyunghyuncho) [inline theano/scan_module/scan.py]: On Mon, Mar 16, 2015 at 4:04 PM, Pascal Lamblin <notifications@github.com>
wrote:

> In theano/scan_module/scan.py
> <https://github.com/Theano/Theano/pull/2622#discussion_r26522029>:
>
> > @@ -314,6 +316,10 @@ def f(x):
> >          Set the value of allow gc for the internal graph of scan.  If
> >          set to None, this will use the value of config.scan.allow_gc.
> >
> > +    :param strict:
> > +        If true, all the shared variables used in ``fn`` must be provided as a
> > +        part of ``non_sequences``.
>
> or sequences.
>
Fixed!

**Comment by kyunghyuncho** [comment]
@nouiz @carriepl @abergeron @lamblin 
Added a unit test which tests if two scan functions (one using strict mode and the other not) return the same result. When run on a local machine, the test passed without any issue.

**Comment by kyunghyuncho** [comment]
@lamblin 

Just added another unit test to test whether MissingInputError is raised when strict=True and non_sequences=[] (or None).
